Remove warning 4 in generated code for sum types

emillon opened the pull request:
### The problem

For a type like:

```ocaml
type t =
| A
| B
| C of int
```

The deriver would generate a single `_` case in the pattern patching for `bin_size` that matches all the constant constructors (`A` and `B`):

```ocaml
let bin_size_t =
  function
  | C v1 -> let size = 1  in Pervasives.(+) size (bin_size_int v1)
  | _ -> 1
```

This triggers an instance of warning 4, "fragile pattern matching". While innocuous in that case, getting rid of it requires it to be globally suppressed.

### The fix

Instead, this replicates the case so that the pattern matching is done extensively:

```ocaml
let bin_size_t =
  function
  | A -> 1
  | B -> 1
  | C v1 -> let size = 1  in Pervasives.(+) size (bin_size_int v1)
```

### Alternative / trade-offs

If the goal is to just remove the warning, it is possible to insert a `[@warning "-4"]` annotation on the generated `match` expression. However, building case per constructor makes it possible to directly use `List.map` without having to do it in two passes, which simplifies the implementation (this refactoring is included). Also, the duplicated branch is small (a single constant), so I'm assuming that the compiled code will be similar in both cases.

### A note on building

I have not managed to build `master`, which I assume is due to versions of dependencies that are not released yet. So I tested this on top of `113.33.03` and blindly rebased on top of `master`.

Let me know if you need anything else.
Thanks!

ghost commented:
Thanks, generating code that doesn't produce warnings is good indeed. Have you considered generating an or-pattern for the constant cases? This should be a more conservative change

ghost commented:
BTW, to build the development branch you can use our development [opam repository](https://github.com/janestreet/opam-repository) which points the the latest dev version of all our packages.

emillon committed:
Remove warning 4

For a type like:

```ocaml
type t =
| A
| B
| C of int
```

The deriver would generate a single `_` case in the pattern patching for
`bin_size`. This triggers an instance of warning 4, "fragile pattern
matching". While innocuous in that case, it requires global suppression
or similar tricks.

Instead, this uses a or-case so that the pattern matching is done
extensively.

emillon commented:
Hi,

Thanks for the pointer to the opam repository. I managed to build the development branch on a fresh 4.03.0 switch (though it fails on 4.04.0 - it seems to be expected).

I just pushed a version that builds a or-pattern instead of `ppat_any`. This uses two passes but is more conservative indeed.

Let me know what you think, and happy new year BTW!

ghost commented:
Sorry for the delay, yep the changes look good. I submitted them for internal review.

About 4.04, yeah it's expected. I'm currently working on detaching our ppx code from the compiler libs to improve the situation.

jeremiedimino committed:
Merge branch 'master' into remove-warning-4

ghost commented:
This has been reviewed and made it's way on github

emillon commented:
Great, thanks! Can't wait to use it from opam :)
